File: scripts/gather-filesizes.py
import json
import os

exclude = set(['resources', 'public', 'node_modules', '.github', '.cache', '.git', '.DS_Store', 'venv', '.venv'])
file_metadata = []

# Taken from https://stackoverflow.com/questions/1094841/get-human-readable-version-of-file-size
def sizeof_fmt(num, suffix='B'):
    for unit in ['','Ki','Mi','Gi','Ti','Pi','Ei','Zi']:
        if abs(num) < 1024.0:
            return "%3.1f%s%s" % (num, unit, suffix)
        num /= 1024.0
    return "%.1f%s%s" % (num, 'Yi', suffix)

for root, dirs, files in os.walk('.', topdown=True):
    dirs[:] = [d for d in dirs if d not in exclude]
    path = root.split(os.sep)
    for file in files:
        full_path = f"{'/'.join(path)}/{file}"
        file_metadata.append({
            'path': full_path,
            'size': os.stat(full_path).st_size,
            'type': os.path.splitext(full_path)[1]
        })

file_metadata = sorted(file_metadata, key = lambda i: i['size'], reverse=True)

for file in file_metadata:
    file['size'] = sizeof_fmt(file['size'])

with open('data/filesizes.json', 'w') as file:
    json.dump(file_metadata, file, indent=2)

File: scripts/gatherFilesizes.js
const fs = require("fs");
const path = require("path");

const exclude = [
  "resources",
  "public",
  "node_modules",
  ".github",
  "cache",
  ".git",
  ".DS_Store",
  "venv",
  ".venv"
]

let file_metadata = []

// Taken from https://stackoverflow.com/a/14919494
function fmtSizeOf(bytes, si = false, dp = 1) {
  const thresh = si ? 1000 : 1024

  if (Math.abs(bytes) < thresh) {
    return bytes + " B"
  }

  const units = si
    ? ["kB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"]
    : ["KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB"]
  let u = -1
  const r = 10 ** dp

  do {
    bytes /= thresh
    ++u
  } while (Math.round(Math.abs(bytes) * r) / r >= thresh && u < units.length - 1)


  return bytes.toFixed(dp) + " " + units[u]
}

async function* walk(dir) {
  if (exclude.includes(path.basename(dir))) return
  for await (const d of await fs.promises.opendir(dir)) {
    const entry = path.join(dir, d.name);
    if (d.isDirectory()) yield* walk(entry);
    else if (d.isFile()) yield entry;
  }
}

async function scanFilesizes() {
  for await (const p of walk(process.cwd())) {
    const file = await fs.promises.stat(p)
    file_metadata.push({
      'created': file.birthtime,
      'path': path.relative(process.cwd(), p),
      'size': fmtSizeOf(file.size),
      'size_nofmt': file.size,
      'type': path.extname(p).split('.').slice(1).join('.')
    })
  }
  file_metadata = file_metadata.sort((a, b) => b.size_nofmt - a.size_nofmt)

  fs.writeFileSync("data/filesizes.json", JSON.stringify(file_metadata, null, 2))
  console.log("~ Filesize data generated")
}

scanFilesizes()